nxontology_ml/model.py

```diff
@@ -0,0 +1,152 @@
+from collections.abc import Iterable
+from dataclasses import dataclass
+from enum import Enum
+from itertools import islice
+from pathlib import Path
+from typing import Any
+
+import numpy as np
+import pandas as pd
+from catboost import CatBoostClassifier, FeaturesData, Pool
+from sklearn.metrics import classification_report
+
+from nxontology_ml.efo import get_efo_otar_slim
+from nxontology_ml.utils import ROOT_DIR
+
+
+@dataclass
+class TrainingRecord:
+    efo_otar_slim_id: str
+    efo_label: str
+    rs_classification: str
+
+
+class FeatureType(Enum):
+    LABEL = 1
+    NUMERICAL = 2
+    CATEGORICAL = 3
+    IGNORED = 4
+
+    @classmethod
+    def from_feature_name(cls, name: str) -> "FeatureType":
+        if name == "__label__":
+            return cls.LABEL
+        if cls._is_numerical(name):
+            return cls.NUMERICAL
+        if cls._is_categorical(name):
+            return cls.CATEGORICAL
+        assert cls._is_ignored(name), f"Unknown feature: {name}"
+        return cls.IGNORED
+
+    @staticmethod
+    def _is_numerical(name: str) -> bool:
+        if name == "depth":
+            return True
+        for prefix in {"n_", "intrinsic_", "xref_"}:
+            if name.startswith(prefix):
+                return True
+        return False
+
+    @staticmethod
+    def _is_categorical(name: str) -> bool:
+        return name in {"prefix", "is_gwas_trait"}
+
+    @staticmethod
+    def _is_ignored(name: str) -> bool:
+        return name in {"identifier", "name"}
+
+
+def read_training_data(
+    take: int | None = None,
+    data_path: Path = ROOT_DIR / "data/efo_otar_slim_v3.43.0_rs_classification.tsv",
+) -> Iterable[dict[str, Any]]:
+    # Get labelled data
+    def _labelled_data() -> Iterable[TrainingRecord]:
+        data = iter(data_path.read_text().splitlines())
+        assert next(data).strip() == "efo_otar_slim_id	efo_label	rs_classification"
+        for line in data:
+            efo_otar_slim_id, efo_label, rs_classification = line.split("\t", 3)
+            yield TrainingRecord(efo_otar_slim_id, efo_label, rs_classification)
+
+    # Get Ontology
+    nxo = get_efo_otar_slim()
+    nodes: set[str] = set(nxo.graph)
+
+    # Filter nodes from the ontology that are labelled
+    for record in islice(_labelled_data(), take):
+        id_ = record.efo_otar_slim_id
+        if id_ in nodes:
+            metrics = nxo.node_info(id_).get_metrics()
+            metrics["__label__"] = record.rs_classification
+            yield metrics
```

Comment:
General commit no action needed.

Note that we can assume that we will never have an ontology where the number of nodes exceeds something that will fit in memory. Therefore, feel liberated to use techniques such as reading into pandas doing a join/merge, etc.

I don't love the pandas API and not insisting we should use it whenever possible, but it's a fair assumption that all collaborators will be familiar with it and therefore it can be a more succinct (sometimes less safe and less disciplined) route to accomplish everything.